website/docs/r/iam_role.html.markdown

```diff
@@ -91,6 +107,128 @@ resource "aws_iam_role" "instance" {
 }
 ```
 
+## Example of Using Exclusive Inline Policies
+
+This example will create an IAM role with two inline IAM policies. If a third policy were added out of band, on the next apply, that policy would be removed. If one of the two original policies were removed, out of band, on the next apply, the policy would be recreated.
+
+```hcl
+resource "aws_iam_role" "example" {
+  name               = "yak_role"
+  assume_role_policy = "${data.aws_iam_policy_document.instance_assume_role_policy.json}" # (not shown)
+
+  inline_policy {
+    name = "my_inline_policy"
+    policy = <<EOF
+{
+  "Statement": [
+    {
+      "Action": [
+        "ec2:Describe*"
+      ],
+      "Effect": "Allow",
+      "Resource": "*"
+    }
+  ],
+  "Version": "2012-10-17"
+}
+EOF
+
+  inline_policy {
+    name_prefix = "inline_"
+    policy = "${data.aws_iam_policy_document.inline_policy.json}"
+  }
+}
+
+data "aws_iam_policy_document" "inline_policy" {
+  statement {
+    actions   = ["ec2:DescribeAccountAttributes"]
+    resources = ["*"]
+  }
+}
+```
+
+## Example of Not Allowing Inline Policies
+
+This example will create an IAM role resource, which, on the next apply, will remove any out-of-band inline policies.
+
+```hcl
+resource "aws_iam_role" "example" {
+  name               = "yak_role"
+  assume_role_policy = "${data.aws_iam_policy_document.instance_assume_role_policy.json}" # (not shown)
+
+  inline_policy {}
+}
+```
+
+## Example of Using Exclusive Managed Policies
+
+This example will create an IAM role resource with two attached managed IAM policies. If a third policy were attached out of band, on the next apply, that policy would be detached. If one of the two original policies were detached out of band, on the next apply, the policy would be recreated and re-attached.
+
+```hcl
+resource "aws_iam_role" "example" {
+  name               = "yak_role"
+  assume_role_policy = "${data.aws_iam_policy_document.instance_assume_role_policy.json}" # (not shown)
+
+  managed_policy_arns = [
+    "${aws_iam_policy.policy_one.arn}",
+    "${aws_iam_policy.policy_two.arn}",
+  ]
+}
+
+resource "aws_iam_policy" "policy_one" {
+  name        = "managed_policy"
+
+  policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Action": [
+        "ec2:Describe*"
+      ],
+      "Effect": "Allow",
+      "Resource": "*"
+    }
+  ]
+}
+EOF
+}
+
+resource "aws_iam_policy" "policy_two" {
+  name        = "managed_policy2"
+
+  policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "s3:ListAllMyBuckets",
+        "s3:ListBucket",
+        "s3:HeadBucket"
+      ],
+      "Resource": "*"
+    }
+  ]
+}
+EOF
+}
+```
+
+## Example of Not Allowing Managed Policies
```

Comment:
Is there also a syntax for specifying that no inline policies should be present, removing any that are?

Reply:
That's the key to this. In the past, we were able to figure out the difference between non-existent (do nothing) and empty (remove any). It's still not a pattern that's commonly used. The language and SDK plugin have all changed a lot in the meantime so fingers crossed we can make it work, and it will be reasonably stable.

Reply:
In places where this kind of exclusive notation is available, the attributes still support assignment, i.e. something like this would indicate empty/remove: `inline_policies = []`. The examples all only show the block-notation for inline_policies, with no assignment. Hence the question...

Reply:
Yes, that's the syntax I'm shoot for based on https://www.terraform.io/docs/language/attr-as-blocks.html. Sets/lists have a nasty habit of appearing to be there with len=0 where declared or not so hopefully get that figured out today.

Reply:
Cool, we're on the same page. I was just a little worried when I didn't see any examples of that syntax!

Reply:
It's work. If you have a chance, look through the docs and test configs (and code, if you want).